components/automate-ui/src/app/modules/infra-proxy/create-environment-modal/create-environment-modal.component.html

```diff
@@ -0,0 +1,110 @@
+              </chef-form-field>
+            </div>
+            <div class="button-bar-box">
+              <chef-button tertiary (click)="closeCreateModal()">Cancel</chef-button>
+              <chef-button [disabled]="!firstFormGroup?.valid || conflictError"  primary id="create-button-object-modal" (click)="createEnvironment()">
+                <chef-loading-spinner *ngIf="creating"></chef-loading-spinner>
+                <span *ngIf="!creating">Create</span>
+                <span *ngIf="creating">Creating Environment...</span>
+              </chef-button>
+            </div>
+          </form>
+        </app-infra-tab>
+        <app-infra-tab tabTitle="Constraints" [active]="constraintsTab" [disabled]="!firstFormGroup?.valid">
+          <div *ngIf="showConstraint">
+            <app-infra-environment-constraint
+              [constraintKeys]="constraintKeysp"
+              (constraintValues)="constraintItemsHandler($event)"
+              [dynamicArray]="dynamicArrayp"
+              [name_id]="name_idp">
+
+            </app-infra-environment-constraint>
+          </div>
+          <div class="button-bar-box">
+            <chef-button tertiary (click)="closeCreateModal()">Cancel</chef-button>
+            <chef-button [disabled]="!constraintArray.length || conflictError"  primary id="create-button-object-modal" (click)="createEnvironment()">
+              <chef-loading-spinner *ngIf="creating"></chef-loading-spinner>
+              <span *ngIf="!creating">Create</span>
+              <span *ngIf="creating">Creating Environment...</span>
+            </chef-button>
+          </div>
+        </app-infra-tab>
+        <app-infra-tab tabTitle="Default Attributes" [active]="defaultTab" [disabled]="!firstFormGroup?.valid">
+          <form [formGroup]="thirdFormGroup">
+            <chef-form-field>
+              <textarea chefInput name="dattr" id="dattr" formControlName="dattr" (change)="onChangeJSON($event)"
+                (keyup)="onChangeJSON($event)" [(ngModel)]="default_attr_value">
+              </textarea>
+              <chef-error *ngIf="dattrParseError">
+                Must be a valid JSON object
+              </chef-error>
+            </chef-form-field>
+            <div class="button-bar-box">
+              <chef-button tertiary (click)="closeCreateModal()">Cancel</chef-button>
+              <chef-button [disabled]="!thirdFormGroup?.valid || conflictError"  primary id="create-button-object-modal" (click)="createEnvironment()">
+                <chef-loading-spinner *ngIf="creating"></chef-loading-spinner>
+                <span *ngIf="!creating">Create</span>
+                <span *ngIf="creating">Creating Environment...</span>
+              </chef-button>
+            </div>
+          </form>
+        </app-infra-tab>
+        <app-infra-tab tabTitle="Override Attributes" [active]="overrideTab" [disabled]="!firstFormGroup?.valid">
+          <form [formGroup]="fourthFormGroup">
```

Comment:
Ah, naming again: and here is where describing the data structure instead of its contents comes back to cause a maintenance issue. 

You have first/second/third/fourth-FormGroups, which describe the position I guess, but why should we care about the position? Rather, what is each form group for? Better names would be DetailsFormGroup, DefaultAttrFormGroup, etc. 

Why did I say it is a maintenance issue? You do not even have a secondFormGroup. 😁

Reply:
Agreed, updated the changes.